### ISSUE TITLE
Should IBControllerServer createSocket() in constructor and throws BindException?

### ISSUE THREAD

**Issue by a8d7e8** [ib-controller/ib-controller#95]
IBControllerServer.createSocket() returns false when the port is being used. Usually this is a signal of the same configuration is running thus the whole startup of TWS / IB gateway should be skipped.

Could createSocket() be invoked from the constructor and let the BindException throws out such that IBController can be notified and respond to such incidence?

I see that the maintainers of this project is very busy replying to support issues. So if the maintainer agree of this approach, I will do the change and submit a PR for review.

Thanks.

PS. I understand that this is not a bug(running the same script without causing problem is not part of the requirement or concern of IBController) but my brain tell me that I want to run the same script over and over again without causing any problem, just to be safe when running in headless mode.

**Comment by rlktradewright** [comment]
I understand your point of view, but I've thought about all this before, and I concluded that the way it works now is best. I'll try to explain my thinking.

First of all, let's imagine you have a nicely automated trading system: it uses IBController to start up early every morning, and it just works, every day without fail, and you don't normally think twice about it.

Now one evening after it's closed down for the day, you decide to do some testing with your paper trading system which you never do while your live system is running, but it just so happens that you haven't changed the IBController server port from its default setting. It gets late and you decide to call it a day, but forget to close down your paper TWS.

Next morning your live system starts up. Which would you prefer to happen?: 

1. The command server detects that the port is already in use and throws a BindException, IBController closes, and TWS is not started. You get up late knowing that your live system always starts reliably every morning, so it takes you a while to notice that in fact it hasn't started.

  or

2. The command server detects that the port is already in use, writes this info to the log, and then carries on. TWS is started as usual, and your live system runs normally (until you try to connect to the command port, at which point the connecting entity gets an error).

Now my preference is definitely for 1. I'd rather that my live trading system is not prevented from running because of something that doesn't actually affect TWS running at all. If I'm one of the very few people who actually use the command server, perhaps from a monitoring application that requests the market data to be refreshed if it thinks something has gone wrong, then that application would detect that it can't connect to IBController and it presumably has a way of notifying bad things to someone. And if I don't use the command server, then it doesn't matter in the slightest if the port is already in use.

Feel free to try and convince me that I've got this all wrong!

**Comment by benalexau** [comment]
@rlktradewright, reading your second last paragraph leads me to believe you prefer option 2, not option 1.

This discussion is really about [fast fail](https://en.wikipedia.org/wiki/Fail-fast) versus [graceful degradation](https://en.wikipedia.org/wiki/Fault_tolerance).

I am highly doubtful the control protocol is a serious problem for most applications, or indeed is even being used. However a bind exception indicates the user has a configuration and/or lifecycle error. What else might be wrong on the machine? Do we really want to expose someone's IB account to financial damage when we know for certain there is something wrong on their machine? I believe a fail-fast exit is the most appropriate response, despite the potential inconvenience. However that inconvenience has been caused by the user's actions (not IBController error, IB error, temporary network error etc) and it could save them a lot of money (eg they never intended the live system to start up, as it was started by a timer/cron job they forgot to disable while in the middle of making untested development changes etc).

The main issue is a fail-fast exit isn't likely to be noticed unless the user has a monitoring approach either via their service manager infrastructure (eg Windows tasks, systemd) or detect it on the automated trading system client side. However everybody building an automated trading system needs to deal with such issues in depth, so they ought to notice and get an email / page / text / whatever.

As an aside, most people don't check service logs once a newly-configured service seems to be working. I'd imagine IBController userse typically modify the INI username and password, spin it up, and never even notice a control port conflict (or if they do they think "that's weird, but it seems to work, so I'll just move on"). If you fail-fast they will notice it immediately and pick a different port number.

**Comment by a8d7e8** [comment]
@rlktradewright Ahha. You've convinced me that my "solution" is over-simplified and not general enough that I haven't consider other people or other cases.

Here was my scenario:

I've set ExistingSessionDetectedAction=primary and mistakenly run two instances without recognizing it as I run it in headless mode. Luckily, I've discovered it before my account has been blocked by too many login attempts.(not sure will IB block me but it has rate limit for historical taking, so .....)

Since headless mode is not part of IBController, maybe the re-entrance (for the identical login id) detection should be done on top of or as a plugin or as a configuration parameter(but it seems that there are already too many configuration parameters) IBC so that the behavior of IBC can be made backward compatible.

Thanks for your explanation.

@benalexau I'm reviewing your package that uses monit. Maybe this is where I should head to. Though I'm not using Arch Linux I will take a look into it. Thanks.

**Comment by rlktradewright** [comment]
@benalexau You're right, it's option 2 for me!

Of course fail fast is the appropriate behaviour when something is going to prevent TWS from running properly, and in most cases that's what IBController does.

But I don't think an incorrect command server port configuration is one of those cases. No-one is going to lose any money as a result of it: TWS will run fine, and so will the trading application. But failing in this case might cause exactly that, loss of money for the user, especially if they're the sort who are prepared to hold positions overnight without a stop loss (my pockets are nothing like deep enough for that kind of strategy with leveraged instruments like futures).

[_Talking of fail fast, one of the most interesting things I've read recently is [this blog post from Joe Duffy](http://joeduffyblog.com/2016/02/07/the-error-model/) about the error model in the C#-like language used in Microsoft Research's experimental Midori operating system. In fact the whole series of posts is fascinating_.]

@a8d7e8 I didn't understand what you were trying to say in your 4th paragraph ('Since headless mode...'). The detection of the attempt to login with the same user id is done by IB's login server, and TWS then displays a dialog to which IBController reacts by pressing the appropriate button. IBController itself is completely unable to determine beforehand whether a repeated login is being attempted.

**Comment by rlktradewright** [comment]
Closed due to lack of further comment.
